### PR TITLE
Fix delete_traces

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/12174?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12174/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12174
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
Fix file_store delete_traces --> we should delete older traces first if max_traces < number of traces with timestamp < max_timestamp_millis

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/store/tracking/file_store.py [hunk 1]
```diff
@@ -1610,24 +1614,28 @@ def delete_traces(
...
                     )
+            trace_info_and_paths.sort(key=lambda x: x[0].timestamp_ms)
+            if max_traces is None:
+                max_traces = len(trace_info_and_paths)
+            for _, trace_path in trace_info_and_paths:
                 if max_traces == 0:
                     break
+                shutil.rmtree(trace_path)
+                deleted_traces += 1
+                max_traces -= 1
```
Comment (harupy) [inline mlflow/store/tracking/file_store.py:1638]: Can we slice `trace_info_and_paths[:max_traces]` and delete them?

File: mlflow/store/tracking/file_store.py [hunk 2]
```diff
@@ -1577,7 +1579,9 @@ def delete_traces(
             experiment_id: ID of the associated experiment.
             max_timestamp_millis: The maximum timestamp in milliseconds since the UNIX epoch for
                 deleting traces. Traces older than or equal to this timestamp will be deleted.
-            max_traces: The maximum number of traces to delete.
+            max_traces: The maximum number of traces to delete. If max_traces is specified, and
+                it is less than the number of traces that would be deleted based on the
+                max_timestamp_millis, the oldest traces will be deleted first.
```
Comment (harupy) [inline mlflow/store/tracking/file_store.py:1584]: Does this align with SQLAlchemy store?

File: mlflow/store/tracking/file_store.py [hunk 3]
```diff
@@ -1610,24 +1614,28 @@ def delete_traces(
...
                     if trace_info.timestamp_ms <= max_timestamp_millis:
-                        shutil.rmtree(trace_path)
-                        deleted_traces += 1
-                        max_traces -= 1
+                        trace_info_and_paths.append((trace_info, trace_path))
                 except MissingConfigException as e:
                     # trap malformed trace exception and log warning
                     request_id = os.path.basename(trace_path)
-                    logging.warning(
+                    _logger.warning(
```
Comment (harupy) [inline mlflow/store/tracking/file_store.py:1626]: looks like we're using `logging.warning`. Can we fix this as a follow-up?

Comment (harupy) [reply mlflow/store/tracking/file_store.py:1626]: or file a good-first issue :)

File: mlflow/store/tracking/file_store.py (hunk 2 above)
Comment (serena-ruan) [reply mlflow/store/tracking/file_store.py:1584]: Yes

File: mlflow/store/tracking/file_store.py (hunk 3 above)
Comment (serena-ruan) [reply mlflow/store/tracking/file_store.py:1626]: Should use _logger everywhere in the file, will file a follow-up

**Review by harupy — APPROVED** [review]
LGTM!

**Review by B-Step62 — CHANGES_REQUESTED** [review]

**Comment by B-Step62** [comment]
LGTM once https://github.com/mlflow/mlflow/pull/12174#discussion_r1620181944 is addressed! Thanks a lot for the fix @serena-ruan!!

File: mlflow/store/tracking/file_store.py [hunk 4]
```diff
@@ -1610,24 +1614,24 @@ def delete_traces(
...
                         f"Malformed trace with request_id '{request_id}'. Detailed error {e}",
-                        exc_info=True,
+                        exc_info=_logger.isEnabledFor(logging.DEBUG),
                     )
-                if max_traces == 0:
-                    break
+            trace_info_and_paths.sort(key=lambda x: x[0].timestamp_ms)
+            deleted_traces = min(len(trace_info_and_paths), max_traces or len(trace_info_and_paths))
```
[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/store/tracking/file_store.py:1631]: We might want to explicitly describe what 0 means in the docstring to eliminate confusion.

[the remark this answers is not in the document]
Comment (B-Step62) [reply mlflow/store/tracking/file_store.py:1631]: yeah, but I think `0` shouldn't mean "delete all":(

[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/store/tracking/file_store.py:1631]: Let's describe this! If max_traces is 0, search_traces does nothing.

[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/store/tracking/file_store.py:1631]: Another option is reject 0.

[the remark this answers is not in the document]
Comment (serena-ruan) [reply mlflow/store/tracking/file_store.py:1631]: But we already reject max_traces <= 0 here https://github.com/mlflow/mlflow/blob/52e59b2717fc035ddad53ceba125397151ddc435/mlflow/store/tracking/file_store.py#L1606

**Review by B-Step62 — APPROVED** [review]
LGTM!

File: mlflow/store/tracking/file_store.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (B-Step62) [reply mlflow/store/tracking/file_store.py:1631]: oops didn't see the validation, stamped, thanks!!

**Comment by serena-ruan** [comment]
Will implement max_traces checks for all stores in a follow-up PR.

**Pull request merged**